File: android/app/src/main/java/com/aitutor/app/data/local/UserPreferences.kt
package com.aitutor.app.data.local

import android.content.Context
import androidx.datastore.preferences.core.booleanPreferencesKey
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.intPreferencesKey
import androidx.datastore.preferences.core.longPreferencesKey
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import com.aitutor.app.domain.model.ExamMode
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import javax.inject.Inject
import javax.inject.Singleton

private val Context.dataStore by preferencesDataStore(name = "user_prefs")

/**
 * Local key/value store for user preferences:
 * - selected exam (IELTS / TOEFL)
 * - onboarding completion
 * - daily goal, streak, gamification state
 */
@Singleton
class UserPreferences @Inject constructor(
    @ApplicationContext private val context: Context
) {
    private object Keys {
        val EXAM_MODE = stringPreferencesKey("exam_mode")
        val ONBOARDING_DONE = booleanPreferencesKey("onboarding_done")
        val DAILY_GOAL_XP = intPreferencesKey("daily_goal_xp")
        val CURRENT_STREAK = intPreferencesKey("current_streak")
        val LAST_PRACTICE_DAY = longPreferencesKey("last_practice_day")
        val HEARTS = intPreferencesKey("hearts")
        val TOTAL_XP = intPreferencesKey("total_xp")
    }

    val examMode: Flow<ExamMode?> = context.dataStore.data.map { prefs ->
        ExamMode.fromName(prefs[Keys.EXAM_MODE])
    }

    val onboardingDone: Flow<Boolean> = context.dataStore.data.map {
        it[Keys.ONBOARDING_DONE] ?: false
    }

    val dailyGoalXp: Flow<Int> = context.dataStore.data.map {
        it[Keys.DAILY_GOAL_XP] ?: DEFAULT_DAILY_GOAL_XP
    }

    val currentStreak: Flow<Int> = context.dataStore.data.map {
        it[Keys.CURRENT_STREAK] ?: 0
    }

    val hearts: Flow<Int> = context.dataStore.data.map {
        it[Keys.HEARTS] ?: MAX_HEARTS
    }

    val totalXp: Flow<Int> = context.dataStore.data.map {
        it[Keys.TOTAL_XP] ?: 0
    }

    suspend fun setExamMode(mode: ExamMode) {
        context.dataStore.edit { it[Keys.EXAM_MODE] = mode.name }
    }

    suspend fun setOnboardingDone(done: Boolean) {
        context.dataStore.edit { it[Keys.ONBOARDING_DONE] = done }
    }

    suspend fun setDailyGoal(xp: Int) {
        context.dataStore.edit { it[Keys.DAILY_GOAL_XP] = xp }
    }

    suspend fun addXp(delta: Int) {
        context.dataStore.edit {
            val current = it[Keys.TOTAL_XP] ?: 0
            it[Keys.TOTAL_XP] = current + delta
        }
    }

    suspend fun loseHeart() {
        context.dataStore.edit {
            val current = it[Keys.HEARTS] ?: MAX_HEARTS
            it[Keys.HEARTS] = (current - 1).coerceAtLeast(0)
        }
    }

    suspend fun refillHearts() {
        context.dataStore.edit { it[Keys.HEARTS] = MAX_HEARTS }
    }

    suspend fun bumpStreakIfNeeded(today: Long) {
        context.dataStore.edit { prefs ->
            val last = prefs[Keys.LAST_PRACTICE_DAY] ?: 0L
            val streak = prefs[Keys.CURRENT_STREAK] ?: 0
            val daysSince = (today - last) / DAY_MILLIS
            prefs[Keys.CURRENT_STREAK] = when {
                last == 0L -> 1
                daysSince == 0L -> streak       // already practiced today
                daysSince == 1L -> streak + 1   // consecutive day
                else -> 1                       // streak broken
            }
            prefs[Keys.LAST_PRACTICE_DAY] = today
        }
    }

    companion object {
        const val MAX_HEARTS = 5
        const val DEFAULT_DAILY_GOAL_XP = 30
        private const val DAY_MILLIS = 24L * 60 * 60 * 1000
    }
}

File: ios/AITutor/Data/Local/UserPreferences.swift
import Foundation
import Combine

/// Local key/value store for user preferences:
/// - selected exam (IELTS / TOEFL)
/// - onboarding completion
/// - daily goal, streak, gamification state
///
/// SwiftUI views observe via `@EnvironmentObject` / `@ObservedObject`. The
/// underlying store is `UserDefaults` — for richer types we'd switch to a
/// file-backed store later.
@MainActor
final class UserPreferences: ObservableObject {
    static let maxHearts = 5
    static let defaultDailyGoalXp = 30
    private static let dayMillis: TimeInterval = 24 * 60 * 60

    private enum Keys {
        static let examMode = "exam_mode"
        static let onboardingDone = "onboarding_done"
        static let dailyGoalXp = "daily_goal_xp"
        static let currentStreak = "current_streak"
        static let lastPracticeDay = "last_practice_day"
        static let hearts = "hearts"
        static let totalXp = "total_xp"
    }

    private let defaults: UserDefaults

    @Published private(set) var examMode: ExamMode?
    @Published private(set) var onboardingDone: Bool
    @Published private(set) var dailyGoalXp: Int
    @Published private(set) var currentStreak: Int
    @Published private(set) var hearts: Int
    @Published private(set) var totalXp: Int

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        self.examMode = ExamMode.fromName(defaults.string(forKey: Keys.examMode))
        self.onboardingDone = defaults.bool(forKey: Keys.onboardingDone)
        self.dailyGoalXp = defaults.object(forKey: Keys.dailyGoalXp) as? Int ?? Self.defaultDailyGoalXp
        self.currentStreak = defaults.integer(forKey: Keys.currentStreak)
        self.hearts = defaults.object(forKey: Keys.hearts) as? Int ?? Self.maxHearts
        self.totalXp = defaults.integer(forKey: Keys.totalXp)
    }

    func setExamMode(_ mode: ExamMode) {
        examMode = mode
        defaults.set(mode.rawValue, forKey: Keys.examMode)
    }

    func setOnboardingDone(_ done: Bool) {
        onboardingDone = done
        defaults.set(done, forKey: Keys.onboardingDone)
    }

    func setDailyGoal(_ xp: Int) {
        dailyGoalXp = xp
        defaults.set(xp, forKey: Keys.dailyGoalXp)
    }

    func addXp(_ delta: Int) {
        totalXp = max(0, totalXp + delta)
        defaults.set(totalXp, forKey: Keys.totalXp)
    }

    func loseHeart() {
        hearts = max(0, hearts - 1)
        defaults.set(hearts, forKey: Keys.hearts)
    }

    func refillHearts() {
        hearts = Self.maxHearts
        defaults.set(hearts, forKey: Keys.hearts)
    }

    /// Bump the streak by one if the user practiced today, or yesterday;
    /// reset to 1 if more than a day has passed.
    func bumpStreakIfNeeded(today: Date = Date()) {
        let lastTimestamp = defaults.double(forKey: Keys.lastPracticeDay)
        let last = lastTimestamp > 0 ? Date(timeIntervalSince1970: lastTimestamp) : nil
        let calendar = Calendar.current
        let todayStart = calendar.startOfDay(for: today)

        let newStreak: Int
        if let last {
            let lastStart = calendar.startOfDay(for: last)
            let days = calendar.dateComponents([.day], from: lastStart, to: todayStart).day ?? 0
            switch days {
            case 0: newStreak = currentStreak               // already practiced today
            case 1: newStreak = currentStreak + 1           // consecutive day
            default: newStreak = 1                          // streak broken
            }
        } else {
            newStreak = 1
        }

        currentStreak = newStreak
        defaults.set(newStreak, forKey: Keys.currentStreak)
        defaults.set(today.timeIntervalSince1970, forKey: Keys.lastPracticeDay)
    }
}
